skills/contextual-embeddings/guide.ipynb

```diff
@@ -0,0 +1,1372 @@
+     "output_type": "stream",
+     "text": [
+      "Processing chunks: 100%|██████████| 737/737 [02:37<00:00,  4.69it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Contextual Vector database loaded and saved. Total chunks processed: 737\n",
+      "Total input tokens without caching: 500383\n",
+      "Total output tokens: 40318\n",
+      "Total input tokens written to cache: 341422\n",
+      "Total input tokens read from cache: 2825073\n",
+      "Total input token savings from prompt caching: 77.04% of all input tokens used were read from cache.\n",
+      "Tokens read from cache come at a 90 percent discount!\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Load the transformed dataset\n",
+    "with open('data/codebase_chunks.json', 'r') as f:\n",
+    "    transformed_dataset = json.load(f)\n",
+    "\n",
+    "# Initialize the ContextualVectorDB\n",
+    "contextual_db = ContextualVectorDB(\"my_contextual_db\")\n",
+    "\n",
+    "# Load and process the data\n",
+    "#note: consider increasing the number of parallel threads to run this faster, or reducing the number of parallel threads if concerned about hitting your API rate limit\n",
+    "contextual_db.load_data(transformed_dataset, parallel_threads=5)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 360,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.53it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@5: 86.37%\n",
+      "Total Score: 0.8637192780337941\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 40.05it/s]\n"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@10: 92.81%\n",
+      "Total Score: 0.9280913978494625\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Evaluating retrieval: 100%|██████████| 248/248 [00:06<00:00, 39.64it/s]"
+     ]
+    },
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Pass@20: 93.78%\n",
+      "Total Score: 0.9378360215053763\n",
+      "Total queries: 248\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "r5 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 5)\n",
+    "r10 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 10)\n",
+    "r20 = evaluate_db(contextual_db, 'data/evaluation_set.jsonl', 20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Contextual BM25\n",
```

Comment:
might be worth having a section before this, maybe a markdown table? that compares the results of the first run with the second